Add support for ffmpeg 5.x to avoid breakage in hardware.cpp

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

### Have you checked that no other similar bug report(s) already exists?

- [X] I have searched and didn't find any similar issues.

### What operating system(s) this bug have occurred on?

Archlinux

### What version(s) of DOSBox-X have this bug?

commit  899f7e6

### Describe the bug

In the coming weeks / months, main distributions will push ffmpeg 5.x either in their testing or main branch.

I tried to build dosbox-x-sdl2 git code with ffmpeg 5.0, and I got this error log while building:

```
hardware.cpp: In function 'void ffmpeg_audio_frame_send()':
hardware.cpp:158:21: error: 'avcodec_encode_audio2' was not declared in this scope; did you mean 'avcodec_encode_subtitle'?
  158 |                 if (avcodec_encode_audio2(ffmpeg_aud_ctx,&pkt,ffmpeg_aud_frame,&gotit) == 0) {
      |                     ^~~~~~~~~~~~~~~~~~~~~
      |                     avcodec_encode_subtitle
hardware.cpp: In function 'void ffmpeg_flush_video()':
hardware.cpp:251:45: error: 'avcodec_encode_video2' was not declared in this scope; did you mean 'avcodec_encode_subtitle'?
  251 |                                         if (avcodec_encode_video2(ffmpeg_vid_ctx,&pkt,NULL,&gotit) == 0) {
      |                                             ^~~~~~~~~~~~~~~~~~~~~
      |                                             avcodec_encode_subtitle
hardware.cpp: In function 'void ffmpeg_flushout()':
hardware.cpp:297:45: error: 'avcodec_encode_audio2' was not declared in this scope; did you mean 'avcodec_encode_subtitle'?
  297 |                                         if (avcodec_encode_audio2(ffmpeg_aud_ctx,&pkt,NULL,&gotit) == 0) {
      |                                             ^~~~~~~~~~~~~~~~~~~~~
      |                                             avcodec_encode_subtitle
hardware.cpp: In function 'void ffmpeg_reopen_video(double, int)':
hardware.cpp:428:45: error: 'AVStream' {aka 'struct AVStream'} has no member named 'codec'
  428 |         ffmpeg_vid_ctx = ffmpeg_vid_stream->codec = avcodec_alloc_context3(ffmpeg_vid_codec);
      |                                             ^~~~~
hardware.cpp:430:9: error: 'avcodec_get_context_defaults3' was not declared in this scope
  430 |         avcodec_get_context_defaults3(ffmpeg_vid_ctx,ffmpeg_vid_codec);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:470:9: error: 'av_frame_set_colorspace' was not declared in this scope
  470 |         av_frame_set_colorspace(ffmpeg_vidrgb_frame,AVCOL_SPC_RGB);
      |         ^~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:479:9: error: 'av_frame_set_color_range' was not declared in this scope
  479 |         av_frame_set_color_range(ffmpeg_vidrgb_frame,AVCOL_RANGE_MPEG);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp: In function 'void CAPTURE_AddImage(Bitu, Bitu, Bitu, Bitu, Bitu, float, uint8_t*, uint8_t*)':
hardware.cpp:1121:64: error: invalid conversion from 'const AVCodec*' to 'AVCodec*' [-fpermissive]
 1121 |                      ffmpeg_aud_codec = avcodec_find_encoder(AV_CODEC_ID_AAC);
      |                                         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
      |                                                             |
      |                                                             const AVCodec*

hardware.cpp:1122:64: error: invalid conversion from 'const AVCodec*' to 'AVCodec*' [-fpermissive]
 1122 |                      ffmpeg_vid_codec = avcodec_find_encoder(AV_CODEC_ID_H264);
      |                                         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
      |                                                             |
      |                                                             const AVCodec*

hardware.cpp:1132:50: error: 'AVFormatContext' {aka 'struct AVFormatContext'} has no member named 'filename'
 1132 |                         snprintf(ffmpeg_fmt_ctx->filename,sizeof(ffmpeg_fmt_ctx->filename),"%s",path.c_str());
      |                                                  ^~~~~~~~
hardware.cpp:1132:82: error: 'AVFormatContext' {aka 'struct AVFormatContext'} has no member named 'filename'
 1132 |    snprintf(ffmpeg_fmt_ctx->filename,sizeof(ffmpeg_fmt_ctx->filename),"%s",path.c_str());
      |                                                             ^~~~~~~~

hardware.cpp:1137:75: error: 'AVFormatContext' {aka 'struct AVFormatContext'} has no member named 'filename'
 1137 |           if (avio_open(&ffmpeg_fmt_ctx->pb,ffmpeg_fmt_ctx->filename,AVIO_FLAG_WRITE) < 0) {
      |                                                             ^~~~~~~~

hardware.cpp:1147:61: error: 'AVStream' {aka 'struct AVStream'} has no member named 'codec'
 1147 |                         ffmpeg_vid_ctx = ffmpeg_vid_stream->codec;
      |                                                             ^~~~~
hardware.cpp:1148:25: error: 'avcodec_get_context_defaults3' was not declared in this scope
 1148 |                         avcodec_get_context_defaults3(ffmpeg_vid_ctx,ffmpeg_vid_codec);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1192:61: error: 'AVStream' {aka 'struct AVStream'} has no member named 'codec'
 1192 |                         ffmpeg_aud_ctx = ffmpeg_aud_stream->codec;
      |                                                             ^~~~~
hardware.cpp:1199:58: error: 'AV_CH_LAYOUT_STEREO' was not declared in this scope
 1199 |                         ffmpeg_aud_ctx->channel_layout = AV_CH_LAYOUT_STEREO;
      |                                                          ^~~~~~~~~~~~~~~~~~~
hardware.cpp:1230:25: error: 'av_frame_set_channels' was not declared in this scope
 1230 |                         av_frame_set_channels(ffmpeg_aud_frame,2);
      |                         ^~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1231:25: error: 'av_frame_set_sample_rate' was not declared in this scope; did you mean 'av_frame_get_side_data'?
 1231 |                         av_frame_set_sample_rate(ffmpeg_aud_frame,(int)capture.video.audiorate);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~
      |                         av_frame_get_side_data
hardware.cpp:1232:25: error: 'av_frame_set_channel_layout' was not declared in this scope
 1232 |                         av_frame_set_channel_layout(ffmpeg_aud_frame,AV_CH_LAYOUT_STEREO);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1240:25: error: 'av_frame_set_colorspace' was not declared in this scope
 1240 |                         av_frame_set_colorspace(ffmpeg_vidrgb_frame,AVCOL_SPC_RGB);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1250:25: error: 'av_frame_set_color_range' was not declared in this scope
 1250 |                         av_frame_set_color_range(ffmpeg_vidrgb_frame,AVCOL_RANGE_MPEG);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1445:37: error: 'avcodec_encode_video2' was not declared in this scope; did you mean 'avcodec_encode_subtitle'?
 1445 |                                 if (avcodec_encode_video2(ffmpeg_vid_ctx,&pkt,ffmpeg_vid_frame,&gotit) == 0) {
      |                                     ^~~~~~~~~~~~~~~~~~~~~
      |                                     avcodec_encode_subtitle

```

### Expected behavior

Build process going on.

### Steps to reproduce the behaviour

1. grab latest code
2. use a distribution providing ffmpeg 5.x
3. try to build dosbox using these options: --enable-core-inline --enable-debug --enable-avcodec --prefix=/usr --enable-sdl2
4. wait and see the crash happening.

### Used configuration

```ini
Build options used: --enable-core-inline --enable-debug --enable-avcodec --prefix=/usr --enable-sdl2

Workaround: replace enable-avcodec by disable-avcodec from command line.
```


### Emulator log

_No response_

### Additional context

_No response_

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,3 +1,6 @@
+Unreleased
+  - Fixed FFMPEG 5 compatibility (Jookia)
+
 2022.09.0 (0.84.3)
   - Updated FFMPEG video capture to use newer API,
     fix memory leaks, and fix use-after-free bug that

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -52,6 +52,7 @@
 #if (C_AVCODEC)
 extern "C" {
 #include <libavutil/pixfmt.h>
+#include <libavutil/channel_layout.h>
 #include <libswscale/swscale.h>
 #include <libavcodec/avcodec.h>
 #include <libavformat/avformat.h>
@@ -72,9 +73,9 @@ bool systemmessagebox(char const * aTitle, char const * aMessage, char const * a
 #if (C_AVCODEC)
 bool ffmpeg_init = false;
 AVFormatContext*	ffmpeg_fmt_ctx = NULL;
-AVCodec*		ffmpeg_vid_codec = NULL;
+const AVCodec*		ffmpeg_vid_codec = NULL;
 AVCodecContext*		ffmpeg_vid_ctx = NULL;
-AVCodec*		ffmpeg_aud_codec = NULL;
+const AVCodec*		ffmpeg_aud_codec = NULL;
 AVCodecContext*		ffmpeg_aud_ctx = NULL;
 AVStream*		ffmpeg_vid_stream = NULL;
 AVStream*		ffmpeg_aud_stream = NULL;
@@ -435,7 +436,7 @@ void ffmpeg_reopen_video(double fps,const int bpp) {
 	LOG_MSG("Restarting video codec");
 
 	// FIXME: This is copypasta! Consolidate!
-	ffmpeg_vid_ctx = ffmpeg_vid_stream->codec = avcodec_alloc_context3(ffmpeg_vid_codec);
+	ffmpeg_vid_ctx = avcodec_alloc_context3(ffmpeg_vid_codec);
 	if (ffmpeg_vid_ctx == NULL) E_Exit("Error: Unable to reopen vid codec");
 	ffmpeg_vid_ctx->bit_rate = 25000000; // TODO: make configuration option!
 	ffmpeg_vid_ctx->keyint_min = 15; // TODO: make configuration option!
@@ -478,7 +479,7 @@ void ffmpeg_reopen_video(double fps,const int bpp) {
 	if (ffmpeg_aud_frame == NULL || ffmpeg_vid_frame == NULL || ffmpeg_vidrgb_frame == NULL)
 		E_Exit(" ");
 
-	av_frame_set_colorspace(ffmpeg_vidrgb_frame,AVCOL_SPC_RGB);
+	ffmpeg_vidrgb_frame->colorspace = AVCOL_SPC_RGB;
 	ffmpeg_vidrgb_frame->width = (int)capture.video.width;
 	ffmpeg_vidrgb_frame->height = (int)capture.video.height;
 	ffmpeg_vidrgb_frame->format = ffmpeg_bpp_pick_rgb_format(bpp);
@@ -486,8 +487,8 @@ void ffmpeg_reopen_video(double fps,const int bpp) {
 		E_Exit(" ");
 	}
 
-	av_frame_set_colorspace(ffmpeg_vid_frame,AVCOL_SPC_SMPTE170M);
-	av_frame_set_color_range(ffmpeg_vidrgb_frame,AVCOL_RANGE_MPEG);
+	ffmpeg_vid_frame->colorspace = AVCOL_SPC_SMPTE170M;
+	ffmpeg_vidrgb_frame->color_range = AVCOL_RANGE_MPEG;
 	ffmpeg_vid_frame->width = (int)capture.video.width;
 	ffmpeg_vid_frame->height = (int)capture.video.height;
 	ffmpeg_vid_frame->format = ffmpeg_vid_ctx->pix_fmt;
@@ -1140,12 +1141,11 @@ skip_shot:
 				LOG_MSG("Failed to allocate format context (mpegts)");
 				goto skip_video;
 			}
-			snprintf(ffmpeg_fmt_ctx->filename,sizeof(ffmpeg_fmt_ctx->filename),"%s",path.c_str());
 
 			if (ffmpeg_fmt_ctx->oformat == NULL)
 				goto skip_video;
 
-			if (avio_open(&ffmpeg_fmt_ctx->pb,ffmpeg_fmt_ctx->filename,AVIO_FLAG_WRITE) < 0) {
+			if (avio_open(&ffmpeg_fmt_ctx->pb,pathvid.c_str(),AVIO_FLAG_WRITE) < 0) {
 				LOG_MSG("Failed to avio_open");
 				goto skip_video;
 			}
@@ -1155,8 +1155,8 @@ skip_shot:
 				LOG_MSG("failed to open audio stream");
 				goto skip_video;
 			}
-			ffmpeg_vid_ctx = ffmpeg_vid_stream->codec;
-
+			ffmpeg_vid_ctx = avcodec_alloc_context3(ffmpeg_vid_codec);
+			if (ffmpeg_vid_ctx == NULL) E_Exit("Error: Unable to open vid context");
 			ffmpeg_vid_ctx->bit_rate = 25000000; // TODO: make configuration option!
 			ffmpeg_vid_ctx->keyint_min = 15; // TODO: make configuration option!
 			ffmpeg_vid_ctx->time_base.num = 1000000;
@@ -1202,7 +1202,8 @@ skip_shot:
 				LOG_MSG("failed to open audio stream");
 				goto skip_video;
 			}
-			ffmpeg_aud_ctx = ffmpeg_aud_stream->codec;
+			ffmpeg_aud_ctx = avcodec_alloc_context3(ffmpeg_aud_codec);
+			if (ffmpeg_aud_ctx == NULL) E_Exit("Error: Unable to open aud context");
 			ffmpeg_aud_ctx->sample_rate = (int)capture.video.audiorate;
 			ffmpeg_aud_ctx->channels = 2;
 			ffmpeg_aud_ctx->flags = 0; // do not use global headers
@@ -1239,9 +1240,9 @@ skip_shot:
 			if (ffmpeg_aud_frame == NULL || ffmpeg_vid_frame == NULL || ffmpeg_vidrgb_frame == NULL)
 				goto skip_video;
 
-			av_frame_set_channels(ffmpeg_aud_frame,2);
-			av_frame_set_sample_rate(ffmpeg_aud_frame,(int)capture.video.audiorate);
-			av_frame_set_channel_layout(ffmpeg_aud_frame,AV_CH_LAYOUT_STEREO);
+			ffmpeg_aud_frame->channels = 2;
+			ffmpeg_aud_frame->sample_rate = (int)capture.video.audiorate;
+			ffmpeg_aud_frame->channel_layout = AV_CH_LAYOUT_STEREO;
 			ffmpeg_aud_frame->nb_samples = ffmpeg_aud_ctx->frame_size;
 			ffmpeg_aud_frame->format = ffmpeg_aud_ctx->sample_fmt;
 			if (av_frame_get_buffer(ffmpeg_aud_frame,16) < 0) {
@@ -1249,7 +1250,7 @@ skip_shot:
 				goto skip_video;
 			}
 
-			av_frame_set_colorspace(ffmpeg_vidrgb_frame,AVCOL_SPC_RGB);
+			ffmpeg_vidrgb_frame->colorspace = AVCOL_SPC_RGB;
 			ffmpeg_vidrgb_frame->width = (int)capture.video.width;
 			ffmpeg_vidrgb_frame->height = (int)capture.video.height;
 			ffmpeg_vidrgb_frame->format = ffmpeg_bpp_pick_rgb_format((int)bpp);
@@ -1258,8 +1259,8 @@ skip_shot:
 				goto skip_video;
 			}
 
-			av_frame_set_colorspace(ffmpeg_vid_frame,AVCOL_SPC_SMPTE170M);
-			av_frame_set_color_range(ffmpeg_vidrgb_frame,AVCOL_RANGE_MPEG);
+			ffmpeg_vid_frame->colorspace = AVCOL_SPC_SMPTE170M;
+			ffmpeg_vidrgb_frame->color_range = AVCOL_RANGE_MPEG;
 			ffmpeg_vid_frame->width = (int)capture.video.width;
 			ffmpeg_vid_frame->height = (int)capture.video.height;
 			ffmpeg_vid_frame->format = ffmpeg_vid_ctx->pix_fmt;

